Set SCHED_BATCH priority on the loadblk thread.

eklitzke opened the pull request:
Today I came across #10271, and while reading the discussion #6358 was linked to. Linux systems have a `SCHED_BATCH` scheduler priority that is useful for threads like loadblk. You can find the full details at [sched(7)](http://man7.org/linux/man-pages/man7/sched.7.html), but I'll quote the relevant part of the man page below:

> ...this policy will cause the scheduler to always assume that the thread is
CPU-intensive. Consequently, the scheduler will apply a small scheduling penalty
with respect to wakeup behavior, so that this thread is mildly disfavored in
scheduling decisions.
>
> This policy is useful for workloads that are noninteractive, but do not want to
lower their nice value, and for workloads that want a deterministic scheduling
policy without interactivity causing extra preemptions (between the workload's
tasks).

I think this change is useful independently of #10271 and irrespective of whether that change is merged. Under normal operation the loadblk thread will just import `mempool.dat`. However, if Bitcoin is started with `-reindex` or `-reindex-chainstate` this thread will use a great deal of CPU while it rebuilds the chainstate database (and the block database in the case of `-reindex`). By setting `SCHED_BATCH` this thread is less likely to interfere with interactive tasks (e.g. the user's web browser, text editor, etc.).

I'm leaving the nice value unchanged (which also affects scheduling decisions) because I think that's better set by the user. Likewise I'm not using [ioprio_set(2)](http://man7.org/linux/man-pages/man2/ioprio_set.2.html) because it can cause the thread to become completely I/O starved (and knowledgeable users can use `ionice(1)` anyway).

eklitzke committed:
Set SCHED_BATCH priority on the loadblk thread.

While reading another PR I saw a mention of #6358. The use case for
SCHED_BATCH is to hint to the kernel that the thread is running a
non-interactive workload that consumes a lot of CPU time. This is
helpful on desktop machines where the loadblk thread can interfere with
interactive applications. More details can be found in the sched(7) man
page.

laanwj commented on src/util.cpp:
```diff
@@ -955,3 +955,10 @@ fs::path AbsPathForConfigVal(const fs::path& path, bool net_specific)
     return fs::absolute(path, GetDataDir(net_specific));
 }
+
+void ScheduleBatchPriority(pid_t tid) {
```
pid_t is not available on windows, I think it would be better to remove the parameter as it's always passed as 0.

eklitzke replied:
I think you're right.

laanwj commented:
Is loadblk thread the only thread this is relevant for? what about the script verification threads?

promag commented:
Is this the windows equivalent?
```cpp
SetThreadPriority(GetCurrentThread(), THREAD_MODE_BACKGROUND_BEGIN)
```

eklitzke commented:
On Linux `SCHED_BATCH` just prevents batch tasks from pre-empting normal tasks (i.e. normal tasks will run until their timeslice expires, or another normal tasks pre-empts them). When picking the next task to run the scheduler considers `SCHED_OTHER` (i.e. normal) and `SCHED_BATCH` tasks equally. This is why setting `SCHED_BATCH` tasks are only "mildly disfavored", because they'll still be considered to run using the normal metrics whenever the scheduler ticks (usually 100Hz or 1000Hz).

It sounds like Windows `THREAD_MODE_BACKGROUND_BEGIN` is [much more extreme](https://stackoverflow.com/questions/43051261/why-does-thread-mode-background-begin-cause-my-code-to-run-20x-slower-than-threa). It looks like maybe the Windows equivalent is `THREAD_PRIORITY_BELOW_NORMAL` but I do not know a lot about Windows.

I don't know as much about the script verification threads. If we are going to set `SCHED_BATCH` for loadblk then it makes sense to me to set the script verification threads to `SCHED_BATCH` as well, at least until the loadblk thread completes. You can freely switch back from `SCHED_BATCH` to `SCHED_OTHER` so it would be possible to only run them in the batch mode while reindexing is happening, or we could just always have them set at batch priority.

JeremyRubin commented:
Concept ACK for this: 

one note: you would want something realtime for scriptcheck threads while they are validating a block incoming from network -- we want these to take priority over other threads aggressively.

eklitzke commented:
Re: Jeremy's comment, there are a few ways to increase the priority of the script verification thread to ensure low-latency block relay. On Linux there exists `SCHED_FIFO` and `SCHED_RR` threads which give threads semi-realtime scheduling guarantees, and can ensure that such threads are always prioritized above other threads. Linux also supports per-thread nice values (this is a Linux extension, POSIX specifies that nice values are per-process).

Using the other scheduling modes is better for realtime behavior, but somewhat dangerous because such threads can pre-empt all other non-realtime processes, meaning that this could be used as a DOS vector. Changing the nice value is a bit safer. I will probably look into this as part of another PR because I'd like to benchmark that change.

MarcoFalke commented:
I think the commits can be squashed?

eklitzke commented:
Thanks @MarcoFalke , squashed to one commit. I'm interested into looking at the RT change Jeremy suggested as well, but that's more impactful so I'd rather do that separately unless people want to see that change done here.

ryanofsky commented on src/init.cpp:
```diff
@@ -627,6 +627,7 @@ void ThreadImport(std::vector<fs::path> vImportFiles)
 {
     const CChainParams& chainparams = Params();
     RenameThread("bitcoin-loadblk");
+    ScheduleBatchPriority();
```
Would be nice to log the error message if sched_batch returns -1. EINVAL / EPERM cases would be good to know about.

ryanofsky commented on src/util.cpp:
```diff
@@ -953,3 +957,12 @@ fs::path AbsPathForConfigVal(const fs::path& path, bool net_specific)
+int ScheduleBatchPriority(void) {
+#if HAVE_SCHED_H && defined(SCHED_BATCH)
+    sched_param param{.sched_priority = 0};
+    return sched_setscheduler(0, SCHED_BATCH, &param);
```
Would calling `pthread_setschedparam` instead of `sched_setscheduler` be a little more portable? It seems like you might not need the configure check using this.

ryanofsky reviewed (commented):
utACK a2ed276b52aec97d37257b7aad73f7f332a926d3

eklitzke commented:
Updated to use `pthread_setschedparam()`. It seems that `<sched.h>` is standard on Posix systems but `SCHED_BATCH` is not, so I updated the `#ifdef` logic a little to reflect that. I also put the error logging in `SchedBatchPriority()` itself as this return 1 on Windows but errno won't be set. Could also do it the other way (return 0 on non-Linux hosts and then move the log statement to the call site).

laanwj commented:
utACK https://github.com/bitcoin/bitcoin/pull/12618/commits/d54874d795a523d7cad5f4a9f4270145e56238f8

MarcoFalke commented:
This appears to give me a segfault:

```
==23339== Thread 16 bitcoin-loadblk:
==23339== Invalid read of size 4
==23339==    at 0x7B5AD38: pthread_setschedparam (in /usr/lib64/libpthread-2.26.so)
==23339==    by 0x492BD6: ScheduleBatchPriority() (util.cpp:1048)
==23339==    by 0x2869B3: ThreadImport(std::vector<boost::filesystem::path, std::allocator<boost::filesystem::path> >) (init.cpp:635)
==23339==    by 0x2A6133: operator()<void (*)(std::vector<boost::filesystem::path>), boost::_bi::list0> (bind.hpp:259)
==23339==    by 0x2A6133: operator() (bind.hpp:1294)
==23339==    by 0x2A6133: boost::detail::thread_data<boost::_bi::bind_t<void, void (*)(std::vector<boost::filesystem::path, std::allocator<boost::filesystem::path> >), boost::_bi::list1<boost::_bi::value<std::vector<boost::filesystem::path, std::allocator<boost::filesystem::path> > > > > >::run() (thread.hpp:116)
==23339==    by 0x54E130C: ??? (in /usr/lib64/libboost_thread.so.1.64.0)
==23339==    by 0x7B5950A: start_thread (in /usr/lib64/libpthread-2.26.so)
==23339==    by 0x962C16E: clone (in /usr/lib64/libc-2.26.so)
==23339==  Address 0x2d0 is not stack'd, malloc'd or (recently) free'd
==23339== 
==23339== 
==23339== Process terminating with default action of signal 11 (SIGSEGV): dumping core
==23339==  Access not within mapped region at address 0x2D0
==23339==    at 0x7B5AD38: pthread_setschedparam (in /usr/lib64/libpthread-2.26.so)
==23339==    by 0x492BD6: ScheduleBatchPriority() (util.cpp:1048)
==23339==    by 0x2869B3: ThreadImport(std::vector<boost::filesystem::path, std::allocator<boost::filesystem::path> >) (init.cpp:635)
==23339==    by 0x2A6133: operator()<void (*)(std::vector<boost::filesystem::path>), boost::_bi::list0> (bind.hpp:259)
==23339==    by 0x2A6133: operator() (bind.hpp:1294)
==23339==    by 0x2A6133: boost::detail::thread_data<boost::_bi::bind_t<void, void (*)(std::vector<boost::filesystem::path, std::allocator<boost::filesystem::path> >), boost::_bi::list1<boost::_bi::value<std::vector<boost::filesystem::path, std::allocator<boost::filesystem::path> > > > > >::run() (thread.hpp:116)
==23339==    by 0x54E130C: ??? (in /usr/lib64/libboost_thread.so.1.64.0)
==23339==    by 0x7B5950A: start_thread (in /usr/lib64/libpthread-2.26.so)
==23339==    by 0x962C16E: clone (in /usr/lib64/libc-2.26.so)
